### ISSUE TITLE
[python-backend] Allow more models

### ISSUE THREAD

**Issue by thakkarparth007** [moyix/fauxpilot#123]
Currently the python backend model only allows choosing from a specific list of models (all codegen models).

Ideally, we should let the users choose any huggingface model that supports text generation.

This will require changes in two places:

1. `setup.sh`/`python_backend()` - allow users to enter any HF model.
2. Verify that we can work with this model
3. `python_backend/model.py` - ensure this file can load and query the models we support (Causal/FIB/EncDec).

Currently `model.py` uses `AutoModelForCausalLM` but we should either use `AutoModel` if it works (last time I tried, it wasn't working for codegen models), or try a few classes of models.

**Comment by moyix** [comment]
Yeah, you're right – `AutoModel` doesn't work unfortunately as it ignores the LM head weights:

```
In [1]: from transformers import AutoModel, AutoTokenizer

In [2]: model = AutoModel.from_pretrained('Salesforce/codegen-350M-mono')
Downloading: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 999/999 [00:00<00:00, 1.53MB/s]
Some weights of the model checkpoint at Salesforce/codegen-350M-mono were not used when initializing CodeGenModel: ['lm_head.weight', 'lm_head.bias']
- This IS expected if you are initializing CodeGenModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing CodeGenModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
```

So maybe we should make `AutoModelForCausalLM` the default and have some way to override that? Unless HF has a list somewhere of the "intended" use for each model? (I would have thought this would just be what `AutoModel` does, but apparently not...)

**Comment by ShisuiUzumaki** [comment]
@moyix Is it a very complex feature to implement? If not, I am interested in taking it up

**Comment by moyix** [comment]
@thakkarparth007 can probably give a more complete answer, but I think it should not be too difficult! The steps outlined in the original issue should work.

If you decide to work on this, please also include a test case that tries to use another (small) model! You can look at https://github.com/fauxpilot/fauxpilot/tree/main/tests/python_backend to see how the current Python tests work.

**Comment by thakkarparth007** [comment]
I think the best way to approach this would be to implement different classes for different model families. 

I can think of 3 families:

1. CausalLMs: models that can be instantiated from AutoModelForCausalLM, e.g, codegen and other gpt-esque models. 
2. Infillers, e.g., incoder. incoder uses AutoModelCausalLM but their generation code is slightly involved? https://huggingface.co/spaces/facebook/incoder-demo/blob/main/modules/app.py )
3. EncDec. e.g., T5

This, and the testing code currently is slightly messy. First the setup code gets run so that the model you're testing gets "installed", and then the model gets tested. The test utility doing the installation is I think hardcoding codegen installation, so you might need to parameterized that. 

Apart from these two don't think there's any other complexity involved.
